File: ai_influencer/src/agents/content_production/mcp.py
import os

from dotenv import load_dotenv
from langchain_mcp_adapters.client import MultiServerMCPClient

load_dotenv()

BUFFER_NOT_ALLOWED_TOOLS = [
    "introspect_schema",
    "execute_query",
    "execute_mutation"
]

async def get_buffer_mcp():
    client = MultiServerMCPClient(
        {
            "buffer": {
                "transport": "http",
                "url": "https://mcp.buffer.com/mcp",
                "headers": {
                    "Authorization": f"Bearer {os.getenv('BUFFER_API_KEY')}"
                }
            }
        }
    )

    buffer_tool = await client.get_tools()
    tools = [t for t in buffer_tool if t.name not in BUFFER_NOT_ALLOWED_TOOLS]
    return tools

File: niche_newsletter/src/orchestrator/mcp.ts
import { MultiServerMCPClient } from "@langchain/mcp-adapters";
import type { DynamicStructuredTool } from "langchain";

const ALPHAXIV_MCP_URL = "https://api.alphaxiv.org/mcp/v1";

const KEEP_ALPHAXIV_TOOLS = new Set([
    "discover_papers",
    "get_paper_content",
    "answer_pdf_queries",
    "read_files_from_github_repository",
    "list_library",
    "save_papers_to_folder",
    "remove_papers_from_folder",
    "move_papers_between_folders",
    "create_folder",
    "rename_folder",
])

export async function getAlphaxivTools(): Promise<DynamicStructuredTool[]> {

    const alphaxivToken = process.env.ALPHAXIV_TOKEN;
    if (!alphaxivToken) {
        console.warn(
            "No api key has been set for Alphaxiv tool"
        );
        return [];
    }

    const client = new MultiServerMCPClient({
        buffer: {
            transport: "http",
            url: ALPHAXIV_MCP_URL,
            headers: {
                Authorization: `Bearer ${alphaxivToken}`
            },
        },
    });

    try {
        const tools = await client.getTools();

        return tools.filter(tool => KEEP_ALPHAXIV_TOOLS.has(tool.name));
    } catch (error) {
        console.error("Failed to load alphaxiv MCP tools:", error);
        return [];
    }
};